src/common/src/util/chunk_coalesce.rs

```diff
@@ -12,14 +12,115 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::cell::Cell;
 use std::iter::FusedIterator;
 use std::mem::swap;
 
 use super::iter_util::ZipEqDebug;
 use crate::array::{ArrayBuilderImpl, ArrayImpl, DataChunk};
+use crate::estimate_size::EstimateSize;
 use crate::row::Row;
 use crate::types::{DataType, ToDatumRef};
 
+trait LimitPolicy {
+    /// The limit policy is reached means that the [`DataChunkBuilder`] should complete a data chunk.
+    fn reach(&self, builder: &DataChunkBuilder) -> bool;
+
+    /// Expected count of the chunk which will be built.
+    ///
+    /// There is no any guarantee between the actual count and the count hint.
+    /// None means it's hard to give a hint.
+    fn count_hint(&self) -> Option<usize> {
+        None
+    }
+}
+
+struct LimitByCount {
+    batch_cnt: usize,
```

Comment:
Better not to name it cnt since cnt usually refers to a counter. Maybe use max_counter/cnt_limit